### PR TITLE
feat: add keycloak auth provider options

### PR DESCRIPTION
**Pull request by nilivingston**
### Description

Adds auth provider options for [Keycloak](https://www.keycloak.org/) to `security/manager.py`. Update example `oauth/config.py` to include "keycloak" and "keycloak_before_17" providers to demonstrate working configurations for Keycloak both before and after its version 17 API changes. Example config changes added to `docs/security.rst` as well.

Working out the necessary parameters for Keycloak integration was time consuming. Hopefully this configuration update and accompanying examples will make this process easier for other users.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Is CRUD MVC related.
- [x] Is Auth, RBAC security related.
- [ ] Changes the security db schema.
- [ ] Introduces new feature
- [ ] Removes existing feature

### CODE REVIEW

File: docs/security.rst [hunk 1]
```diff
@@ -233,13 +233,43 @@ Specify a list of OAUTH_PROVIDERS in **config.py** that you want to allow for yo
...
+                "client_id": "KEYCLOAK_CLIENT_ID",
+                "client_secret": "KEYCLOAK_CLIENT_SECRET",   
+                "api_base_url": "https://KEYCLOAK_DOMAIN/auth/realms/master/protocol/openid-connect",
+                "client_kwargs": {
+                    "scope": "email profile"
+                },
+                "access_token_url": "KEYCLOAK_DOMAIN/auth/realms/master/protocol/openid-connect/token",
+                "authorize_url": "KEYCLOAK_DOMAIN/auth/realms/master/protocol/openid-connect/auth",
+                "request_token_url": None,
+        },
```
Comment (dpgaspar) [inline docs/security.rst:265]: missing `}`

File: docs/security.rst [hunk 2]
```diff
@@ -233,13 +233,43 @@ Specify a list of OAUTH_PROVIDERS in **config.py** that you want to allow for yo
...
+            "remote_app": {
+                "client_id": "KEYCLOAK_CLIENT_ID",
+                "client_secret": "KEYCLOAK_CLIENT_SECRET",   
+                "api_base_url": "https://KEYCLOAK_DOMAIN/realms/master/protocol/openid-connect",
+                "client_kwargs": {
+                    "scope": "email profile"
+                },
+                "access_token_url": "KEYCLOAK_DOMAIN/realms/master/protocol/openid-connect/token",
+                "authorize_url": "KEYCLOAK_DOMAIN/realms/master/protocol/openid-connect/auth",
+                "request_token_url": None,
```
Comment (dpgaspar) [inline docs/security.rst:249]: missing `}`

**Review by dpgaspar — COMMENTED** [review]
Looks good, just the comment on the docs.

We are really needing tests for all these providers

File: docs/security.rst (hunk 1 above)
Comment (nilivingston) [reply docs/security.rst:267]: Added the brace.

File: docs/security.rst (hunk 2 above)
Comment (nilivingston) [reply docs/security.rst:249]: Added brace.

**Comment by nilivingston** [comment]
I tested the Keycloak integration manually using the Bitnami Keycloak Docker image. I'm not sure how best to develop integration tests for Keycloak and the other providers for inclusion into the automated test suite. I'll give it some thought.

**Comment by dpgaspar** [comment]
> I tested the Keycloak integration manually using the Bitnami Keycloak Docker image. I'm not sure how best to develop integration tests for Keycloak and the other providers for inclusion into the automated test suite. I'll give it some thought.

That would be great, thank you once more

**Review by dpgaspar — APPROVED** [review]

**Comment by dpgaspar** [comment]
@nilivingston conflict here, I merged another docs related PR. Can you fix it plz?

**Comment by nilivingston** [comment]
> @nilivingston conflict here, I merged another docs related PR. Can you fix it plz?

I resolved the conflict by including both the keycloak and azure examples in `security.rst`. Should be good to go.

**Pull request merged**